[parachute] Parachute microservice document (#3)
Skydio/mavlink-devguide: en/services/parachute.md

# Parachute Protocol

## Introduction

The parachute protocol allows MAVLink control over the behavior of a parachute that is integrated or attached to the drone. Along with this, the parachute protocol also publishes status information for developers or users. 

## MAVLink Parachute Implementations

These parachutes have built-in MAVLink support:

> - AVSS Parachute for X10

Reviewer: AVSS Parachute for Skydio X10?

## Message/Enum Summary

| Message                                                                                                              | Description               |
| ------------------------------------------------------------------------------------------------------------------ | ------------------------- |

> | <span id="PARACHUTE_STATUS"></span>[PARACHUTE_STATUS](../messages/common.md#PARACHUTE_STATUS)                                           | Current status of the parachute. Recommended to publish this at a regular rate.  

Reviewer 2: is it recommended to be published at a regular rate, or is that a requirement? I think it's necessary, at least for our system. Also, would be good to specify what the nominal rate should be. Do they send this at 1Hz right now?

Author: added nominal rate but i think its recommended but required for ours

| <span id="PARACHUTE_TRIGGER_FLAGS"></span>[PARACHUTE_TRIGGER_FLAGS](../messages/common.md#PARACHUTE_TRIGGER_FLAGS) | Parachute trigger sources.        |
| <span id="PARACHUTE_DEPLOYMENT_TRIGGER"></span>[PARACHUTE_DEPLOYMENT_TRIGGER](../messages/common.md#PARACHUTE_DEPLOYMENT_TRIGGER) | Parachute deployment trigger source. |
| <span id="PARACHUTE_SAFETY_FLAGS"></span>[PARACHUTE_SAFETY_FLAGS](../messages/common.md#PARACHUTE_SAFETY_FLAGS) | Parachute module safety-related flags. |
| <span id="PARACHUTE_ERROR_FLAGS"></span>[PARACHUTE_ERROR_FLAGS](../messages/common.md#PARACHUTE_ERROR_FLAGS) | Fault/health indications. |

## Implementation and Messages

### Parachute Connection

Parachutes are expected to follow the [Heartbeat/Connection Protocol](https://github.com/mavlink/mavlink-devguide/blob/master/en/services/heartbeat.md) and send a constant flow of heartbeats (nominally at 1Hz). Parachutes are identified via their type [MAV_TYPE_PARACHUTE](#MAV_TYPE_PARACHUTE).
Individual parachutes are distinguished via their unique component ID, which by default should be [MAV_COMP_ID_PARACHUTE](#MAV_COMP_ID_PARACHUTE) (though this is not mandated and any ID may be used).

> Once a heartbeat is received, the drone can then send a [MAV_CMD_REQUEST_MESSAGE](https://mavlink.io/en/messages/common.html#MAV_CMD_REQUEST_MESSAGE) command to the parachute to receive information about its status. [MAV_CMD_SET_PARACHUTE_ARM](#MAV_CMD_SET_PARACHUTE_ARM) can also then be sent to configure the arming status of the various possible trigger mechanisms of the parachute. 

Reviewer: since `PARACHUTE_STATUS` should be publishing at default rate by default, is the `MAV_CMD_REQUEST_MESSAGE` referring to getting info like software version from `COMPONENT_INFORMATION_BASIC`?

Author: I was referencing in general you can request any message related to parachute. Even though `PARACHUTE_STATUS` is publishing at a default rate, a user could request  the status or `COMPONENT_INFORMATION_BASIC` I will update so that part is more clear though

|--------------------------|------------------------------------------------------------------------------------------------------------------------------------------------|
| `time_boot_ms`           | Time since system boot up in milliseconds.                                                                                                     |
| `error_status`           | Bitmask detailing the [PARACHUTE_ERROR_FLAGS](#PARACHUTE_ERROR_FLAGS).                                                                         |
| `arm_status`             | Bitmask detailing the arming status via [PARACHUTE_TRIGGER_FLAGS](#PARACHUTE_TRIGGER_FLAGS).                                                   |
| `deployment_status`      | Parachute deployment status.                                                                                                                   |
| `safety_status`          | Parachute safety status.                                                                                                                       |

> | `ats_arm_altitude`       | Parachute Automatic Trigger System (ATS) auto-arming/disarming altitude in meters.                                                             |

Reviewer 2: what is this? Is this the MDA? I don't think the calling it the "auto-arming/disarming altitude" is accurate?

Author: This is not used in our product but is something in the protocol we are upstreaming.

#### PARACHUTE_ERROR_FLAGS
[PARACHUTE_ERROR_FLAGS](#PARACHUTE_ERROR_FLAGS) is a bitmask that details the various critical errors that may occur within the parachute module or within the communication between the parachute and the drone. This is published in the [PARACHUTE_STATUS](#PARACHUTE_STATUS) message.

A list of the various errors can be found below:
| Value    | Parameter                                    | Description                                                                                                     |
|----------|----------------------------------------------|-----------------------------------------------------------------------------------------------------------------|
| 1        | `PARACHUTE_ERROR_FLAGS_BAROMETER_ERROR`      | There is an error with the parachute barometer.                                                                  |
| 2        | `PARACHUTE_ERROR_FLAGS_IMU_ERROR`            | There is an error with the parachute IMU.                                                                        |

> | 4        | `PARACHUTE_ERROR_FLAGS_RF_CONNECTION_ERROR`  | There is an error with the parachute's RF that is used for manual control.                                       |

Reviewer 2: "parachute's RF connection"

> | 8        | `PARACHUTE_ERROR_FLAGS_LOW_POWER`            | Parachute module has low power.                                                                                  |

Reviewer 2: what does "low power" mean?

Author: updated with more info

| 32       | `PARACHUTE_ERROR_FLAGS_EFTS_CONNECTION_ERROR`| There is an error with the connection between parachute and Electrical Flight Termination System (EFTS).         |
| 64       | `PARACHUTE_ERROR_FLAGS_POD_CONNECTION_ERROR` | There is an error with the parachute pod.                                                                        |
| 128      | `PARACHUTE_ERROR_FLAGS_EFTS_DIAGNOSE`        | Parachute Electrical Flight Termination System (EFTS) diagnosis failed.                                          |
| 256      | `PARACHUTE_ERROR_FLAGS_CHARGING_FAILED`      | Parachute module charging failed.                                                                                |
| 512      | `PARACHUTE_ERROR_FLAGS_EXTERNAL_POWER_ERROR` | There is an error with the parachute external power source.                                                      |
| 1024     | `PARACHUTE_ERROR_FLAGS_GS_CONNECTION_ERROR`  | There is an error with the connection between parachute and Ground Station (GS).                                 |

> | 2048     | `PARACHUTE_ERROR_FLAGS_GPS_ERROR`            | There is an error with the parachute's GPS.                                                                      |

Reviewer 2: does the parachute have a GPS? Do they have other variants that have a GPS?

Reviewer 2: Guessing they do, and they use it for geofencing? Is that what we plan to do for EU?

> | 4096     | `PARACHUTE_ERROR_FLAGS_SUBSYSTEM_CONNECTION_ERROR` | There is an error with the connection between parachute and subsystem (e.g. remote controller, expansion board, etc.). |

Reviewer 2: what is a subsystem in this case?

| 8192     | `PARACHUTE_ERROR_FLAGS_SUBSYSTEM_FW_ERROR`   | There is an error with the parachute subsystem firmware (e.g. wrong firmware version).                           |
| 16384    | `PARACHUTE_ERROR_FLAGS_RESERVED_1`           | Reserved for future use.                                                                                         |
| 32768    | `PARACHUTE_ERROR_FLAGS_RESERVED_2`           | Reserved for future use.                                                                                         |
| 65536    | `PARACHUTE_ERROR_FLAGS_LOGGING_ERROR`        | There is an error with the parachute's internal logging system.                                                  |
| 131072   | `PARACHUTE_ERROR_FLAGS_MODULE_RETIRED`       | This parachute module is retired (i.e. too many deployments).                                                    |
| 262144   | `PARACHUTE_ERROR_FLAGS_GLOW_WIRE_ERROR`      | There is an error with the parachute glow wire.                                                                  |

> | 524288   | `PARACHUTE_ERROR_FLAGS_OFFBOARD_CONNECTION_ERROR` | There is an error with the MAVLink connection between parachute and offboard computer.                        |

Reviewer 2: what's the difference between offboard and FC? Also, this is the only one that describes the connection as a "MAVLink connection"

> | 1048576  | `PARACHUTE_ERROR_FLAGS_IMU_CALIBRATION_ERROR`| Parachute's internal IMU calibration failed.                                                                     |

Reviewer 2: why do we have `IMU_ERROR` and `IMU_CALIBRATION_ERROR`? What does `IMU_ERROR` actually mean?

Author: updated with distinction

#### MAV_CMD_SET_PARACHUTE_ARM
This command can be used to arm/disarm the various parachute trigger sources. By configuring the arm/disarm setting of the various trigger sources, a user is able to change the behavior of the parachute when it is being used. The operation follows the normal [Command Protocol](https://github.com/mavlink/mavlink-devguide/blob/master/en/services/command.md) rules for command/acknowledgment.

> There are two parameters for this command. The first parameter is used to indicate the arm/disarm setting. Setting a bit to "0" indicated disarm and setting a bit to "1" indicates arm. The second parameter is a mask-byte that indicates which trigger source is being configured. The bits of the second parameter are mapped to [PARACHUTE_TRIGGER_FLAGS](#PARACHUTE_TRIGGER_FLAGS). The second parameter allows users to not have to track the status of all trigger sources when you want to set arm/disarm for specific sources. 

Reviewer: 1. Setting a bit to "0" ~~indicated~~ _indicates_ disarm
2. "The bits of the second parameter are mapped to `PARACHUTE_TRIGGER_FLAGS`" -> both parameters use `PARACHUTE_TRIGGER_FLAGS`

Reviewer 2: also, mask-byte -> bitmask. And the last sentence reads kinda weird. Suggestion:
"The parachute will ignore all bit positions in the first parameter that are set to zero in the second parameter, which allows the user to granularly arm/disarm specific sources individually."

For example, the user wants to arm the automatic trigger system and flight controller trigger, and the user wants to disarm the manual and geofence trigger. 

> - The first parameter would be `0b01100`.

Reviewer: `0bxxx0x110`?

   - Bit 0 for `PARACHUTE_TRIGGER_FLAGS_MANUAL` is set to "0" to disarm.
   - Bit 1 for `PARACHUTE_TRIGGER_FLAGS_ATS` is set to "1" to arm.
   - Bit 2 for `PARACHUTE_TRIGGER_FLAGS_FC` is set to "1" to arm.
   - Bit 4 for `PARACHUTE_TRIGGER_FLAGS_GEOFENCE` is set to "0" to disarm.

> - The second parameter would be `0b11001`.

Reviewer: `0b00010111`?

   - Since bit 0, 1, 2, and 4 are being configured and no other trigger sources are being configured, all of those bits are set to "1" to indicate that those specific trigger sources need their arm statuses to be updated.

>    - Although bit 3 in the first parameter is set to "0", since bit 3 is set to "0" in the second parameter, the parachute should ignore the bit in the first parameter and not change the arm/disarm status of the trigger source mapped to bit 3.

Reviewer: ```suggestion
   - Doesn't matter what bit 3 in the first parameter is set to , since bit 3 is set to "0" in the second parameter, the parachute should ignore the bit in the first parameter and not change the arm/disarm status of the trigger source mapped to bit 3.
```

Reviewer 2: we could also delete this, or say something like "all other bits, such as bit 3, are ignored"

##### PARACHUTE_TRIGGER_FLAGS

[PARACHUTE_TRIGGER_FLAGS](#PARACHUTE_TRIGGER_FLAGS) is a bitmask that is used in the parameters of the [MAV_CMD_SET_PARACHUTE_ARM](#MAV_CMD_SET_PARACHUTE_ARM) command and is published in the [PARACHUTE_STATUS](#PARACHUTE_STATUS) message. When published in the [PARACHUTE_STATUS](#PARACHUTE_STATUS) message, the bitmask details the various arm/disarm states of all the trigger flags.

A list of the various flags can be found below:

> | Bit| Value  | Parameter                                      | Description                                                                                                            |

Reviewer 2: is it standard for mavlink docs to have a "bit" and a "value" colunm for bitmask values? If so, we should add a "bit" column above as well

Author: done!

| 1  | 2      | `PARACHUTE_TRIGGER_FLAGS_ATS`                 | Automatic trigger system (ATS)                                                                                         |
| 2  | 4      | `PARACHUTE_TRIGGER_FLAGS_FC`                  | Flight controller trigger (e.g. MAVLink from FC, PWM, DroneCan, RC Control)                                            |
| 3  | 8      | `PARACHUTE_TRIGGER_FLAGS_OFFBOARD`            | Offboard computer trigger (via MAVLink)                                                                                |
| 4  | 16     | `PARACHUTE_TRIGGER_FLAGS_GEOFENCE`            | Geofence trigger (by parachute). Parachute uses MAVLink mission protocol to fetch geofence.                            |
| 5  | 32     | `PARACHUTE_TRIGGER_FLAGS_FTS_PRECHECKING`     | FTS (flight termination system) pre-checking protocol trigger                                                          |
| 6  | 64     | `PARACHUTE_TRIGGER_FLAGS_ATS_AUTO_ARM`        | Auto-arming of parachute automatic trigger system (ATS). This allows a parachute to enable ATS after reaching a desired altitude. |

> | 7  | 128    | `PARACHUTE_TRIGGER_FLAGS_ATS_AUTO_DISARM`     | Auto-disarming of parachute automatic trigger system (ATS). This allows a parachute to disable ATS after detecting that it is below a desired altitude. |

Reviewer 2: I thought auto-disarm is not altitude based, but stationary upon landing based?

Author: dependent on the developer. I think this can be kept as is to mirror the PR.

Reviewer 2: we should then specify that the implementation of the auto arm/disarm conditions is developer-specific

| 3     | `PARACHUTE_DEPLOYMENT_TRIGGER_DRONE`               | Drone                                                                   |
| 4     | `PARACHUTE_DEPLOYMENT_TRIGGER_MAVLINK`             | MAVLink, e.g. from an offboard computer                                 |
| 5     | `PARACHUTE_DEPLOYMENT_TRIGGER_GEOFENCE`            | Geofence                                                                |
| 6     | `PARACHUTE_DEPLOYMENT_TRIGGER_FTS_PRECHECKING`     | Flight Termination System (FTS) pre-checking protocol                   |

#### PARACHUTE_SAFETY_FLAGS
[PARACHUTE_SAFETY_FLAGS](#PARACHUTE_SAFETY_FLAGS) is a bitmask that details the safety status of the parachute. This is published in the [PARACHUTE_STATUS](#PARACHUTE_STATUS) message.

A breakdown of the bitmask can be found below:

| Value | Parameter                                             | Description                                                                                                     |
|-------|-------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------|

> | 1     | `PARACHUTE_SAFETY_FLAGS_GROUND_CLEARED`               | The parachute module has cleared a safe distance from the ground for deployment.  |

Reviewer 2: if I was a user reading this documentation, my first question would be "what does safe distance mean? Does this mean that I can't trigger the parachute through any means if this is not set, even if it's armed?"

Author: added additional info but it is custom set per parachute

| 2     | `PARACHUTE_SAFETY_FLAGS_ON_GROUND`                    | The parachute's own sensor has confirmed it is stably on the ground.              |
| 4     | `PARACHUTE_SAFETY_FLAGS_GEOFENCE_MISSION_SET`         | The parachute module has downloaded geofence mission successfully and can be triggered by geofence source. |

## Test Script

#### Description

The test suite included in `parachute.py` allows for testing both sides of the parachute communication interaction. The first script `parachute.py` will emulate a standard parachute module. The second script will run a standard test suite against the emulator, testing all commands listed in this document.

##### Instructions

> 1. Run simple parachute emulator `python3 parachute.py`

Reviewer 2: nit: Run the parachute emulator

| 15  | 32768    | `PARACHUTE_ERROR_FLAGS_RESERVED_2`             | Reserved for future use.                                                                                                   |
| 16  | 65536    | `PARACHUTE_ERROR_FLAGS_LOGGING_ERROR`          | There is an error with the parachute's internal logging system.                                                            |
| 17  | 131072   | `PARACHUTE_ERROR_FLAGS_MODULE_RETIRED`         | This parachute module is retired (i.e. too many deployments).                                                              |
| 18  | 262144   | `PARACHUTE_ERROR_FLAGS_GLOW_WIRE_ERROR`        | There is an error with the parachute glow wire.                                                                            |
| 19  | 524288   | `PARACHUTE_ERROR_FLAGS_OFFBOARD_CONNECTION_ERROR` | There is an error with the MAVLink connection between parachute and offboard computer.                                 |
| 20  | 1048576  | `PARACHUTE_ERROR_FLAGS_IMU_CALIBRATION_ERROR`  | Parachute's internal IMU calibration failed.                                                                               |

>                                                                     |

Reviewer: this looks to be extra and not actually part of the table

Author: yes good catch i forgot to double check the preview

[PARACHUTE_ERROR_FLAGS_IMU_ERROR](#PARACHUTE_ERROR_FLAGS_IMU_ERROR) is typically used to indicate a hardware failure of the onboard IMU, and [PARACHUTE_ERROR_FLAGS_IMU_CALIBRATION_ERROR](#PARACHUTE_ERROR_FLAGS_IMU_CALIBRATION_ERROR) may indicate that IMU has no calibration or needs re-calibration.

Depending on the parachute product, not all errors may be applicable. For example, [PARACHUTE_ERROR_FLAGS_LOW_POWER](#PARACHUTE_ERROR_FLAGS_LOW_POWER) would be used if there was an additional power source within the parachute module itself that has its power level monitored. 

> #### MAV_CMD_SET_PARACHUTE_ARM

Reviewer: ```suggestion
### MAV_CMD_SET_PARACHUTE_ARM
```

Reviewer: Oh I see are these all things that can contribute to the different fields of `PARACHUTE_STATUS`? `MAV_CMD_SET_PARACHUTE_ARM` seems more to be on the same level as `PARACHUTE_STATUS` while `PARACHUTE_ERROR_FLAGS`, `PARACHUTE_TRIGGER_FLAGS`, `PARACHUTE_DEPLOYMENT_TRIGGER`, and `PARACHUTE_SAFETY_FLAGS` are on a separate level. But however you want to structure this, either way works

Author: you are correct and i think ill just update this one!!! and change the structure

For example, the user wants to arm the automatic trigger system and flight controller trigger, and the user wants to disarm the manual and geofence trigger. 
- The first parameter would be `0bxxx0x110`.
   - Bit 0 for `PARACHUTE_TRIGGER_FLAGS_MANUAL` is set to "0" to disarm.
   - Bit 1 for `PARACHUTE_TRIGGER_FLAGS_ATS` is set to "1" to arm.
   - Bit 2 for `PARACHUTE_TRIGGER_FLAGS_FC` is set to "1" to arm.
   - Bit 4 for `PARACHUTE_TRIGGER_FLAGS_GEOFENCE` is set to "0" to disarm.
- The second parameter would be `0b00010111`.
   - Since bit 0, 1, 2, and 4 are being configured and no other trigger sources are being configured, all of those bits are set to "1" to indicate that those specific trigger sources need their arm statuses to be updated.
   - All other bits, such as bit 3, are ignored. The parachute should not change the arm/disarm status of the trigger source mapped to bit 3.

> ##### PARACHUTE_TRIGGER_FLAGS

Reviewer: ```suggestion
#### PARACHUTE_TRIGGER_FLAGS
```

| 1     | `PARACHUTE_DEPLOYMENT_TRIGGER_MANUAL`              | Manual                                                                  |
| 2     | `PARACHUTE_DEPLOYMENT_TRIGGER_ATS`                 | Automatic trigger system (ATS)                                          |
| 3     | `PARACHUTE_DEPLOYMENT_TRIGGER_DRONE`               | Drone                                                                   |
| 4     | `PARACHUTE_DEPLOYMENT_TRIGGER_MAVLINK`             | MAVLink, e.g. from an offboard computer                                 |
| 5     | `PARACHUTE_DEPLOYMENT_TRIGGER_GEOFENCE`            | Geofence                                                                |
| 6     | `PARACHUTE_DEPLOYMENT_TRIGGER_FTS_PRECHECKING`     | Flight Termination System (FTS) pre-checking protocol                   |

#### PARACHUTE_SAFETY_FLAGS
[PARACHUTE_SAFETY_FLAGS](#PARACHUTE_SAFETY_FLAGS) is a bitmask that details the safety status of the parachute. This is published in the [PARACHUTE_STATUS](#PARACHUTE_STATUS) message.

A breakdown of the bitmask can be found below:

> | Value | Parameter                                             | Description                                                                                                     |

Reviewer: nit: can also add `Bit` column here to keep it consistent with `PARACHUTE_ERROR_FLAGS` and `PARACHUTE_TRIGGER_FLAGS`

| 2  | 4      | `PARACHUTE_TRIGGER_FLAGS_FC`                  | Flight controller trigger (e.g. MAVLink from FC, PWM, DroneCan, RC Control)                                            |
| 3  | 8      | `PARACHUTE_TRIGGER_FLAGS_OFFBOARD`            | Offboard computer trigger (via MAVLink)                                                                                |
| 4  | 16     | `PARACHUTE_TRIGGER_FLAGS_GEOFENCE`            | Geofence trigger (by parachute). Parachute uses MAVLink mission protocol to fetch geofence.                            |
| 5  | 32     | `PARACHUTE_TRIGGER_FLAGS_FTS_PRECHECKING`     | FTS (flight termination system) pre-checking protocol trigger                                                          |
| 6  | 64     | `PARACHUTE_TRIGGER_FLAGS_ATS_AUTO_ARM`        | Auto-arming of parachute automatic trigger system (ATS). This allows a parachute to enable ATS after reaching a desired altitude. |
| 7  | 128    | `PARACHUTE_TRIGGER_FLAGS_ATS_AUTO_DISARM`     | Auto-disarming of parachute automatic trigger system (ATS). This allows a parachute to disable ATS after detecting that it is below a desired altitude. |

## Test Script

#### Description

> The test suite included in `parachute.py` allows for testing both sides of the parachute communication interaction. The first script `parachute.py` will emulate a standard parachute module. The second script will run a standard test suite against the emulator, testing all commands listed in this document.

Reviewer 2: the test suite is included in `test_parachute.py`, not `parachute.py`
